### PR TITLE
Handle properly closing the WS connection

### LINKED ISSUE #1057: RPC Client: No closing handshake
The RPC client does not close the correction correctly.

### Sample Command
`nimiq-rpc block-number`

### Server Log
`ERROR nimiq_jsonrpc_server | Websocket error: HTTP error: WebSocket protocol error: Connection reset without closing handshake `

### PR DESCRIPTION
**Pull request by viquezclaudio**
Closes #1057

## What's in this pull request?

...
## Pull request checklist

- [ ] All tests pass. The project builds and runs.
- [ ] I have resolved any merge conflicts.
- [ ] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

**Comment by viquezclaudio** [comment]
Requires:
https://github.com/nimiq/jsonrpc/pull/12

File: rpc-client/src/main.rs [hunk 1]
```diff
@@ -102,9 +106,14 @@ impl Client {
             wallet: WalletProxy::new(client.clone()),
             validator: ValidatorProxy::new(client.clone()),
             network: NetworkProxy::new(client.clone()),
-            zkp_component: ZKPComponentProxy::new(client),
+            zkp_component: ZKPComponentProxy::new(client.clone()),
+            ws_client: client,
         })
     }
+
+    pub async fn close(&mut self) {
```
Comment (jsdanielh) [inline rpc-client/src/main.rs:112]: Add some `rustdoc` documentation

File: rpc-client/src/subcommands/accounts_subcommands.rs [hunk 2]
```diff
@@ -1,16 +1,18 @@
+use std::sync::Arc;
+
 use anyhow::Error;
 use async_trait::async_trait;
-
 use clap::Parser;
-use nimiq_rpc_interface::{blockchain::BlockchainInterface, wallet::WalletInterface};
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/accounts_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/blockchain_subcommands.rs [hunk 3]
```diff
@@ -1,16 +1,19 @@
+use std::sync::Arc;
+
 use anyhow::Error;
 use async_trait::async_trait;
 use clap::ArgGroup;
 use clap::Parser;
 use futures::StreamExt;
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/blockchain_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/mempool_subcommands.rs [hunk 4]
```diff
@@ -1,6 +1,10 @@
+use std::sync::Arc;
+
 use anyhow::Error;
 use async_trait::async_trait;
 use clap::Parser;
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/mempool_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/network_subcommands.rs [hunk 5]
```diff
@@ -1,11 +1,14 @@
+use std::sync::Arc;
+
 use anyhow::Error;
 use async_trait::async_trait;
 use clap::Parser;
-use nimiq_rpc_interface::network::NetworkInterface;
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/network_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/policy_subcommands.rs [hunk 6]
```diff
@@ -1,12 +1,14 @@
-use clap::Parser;
+use std::sync::Arc;
 
 use anyhow::Error;
 use async_trait::async_trait;
-use nimiq_rpc_interface::policy::PolicyInterface;
+use clap::Parser;
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/policy_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/transactions_subcommands.rs [hunk 7]
```diff
@@ -1,4 +1,10 @@
+use std::sync::Arc;
+
+use anyhow::Error;
+use async_trait::async_trait;
 use clap::{Args, Parser};
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/transactions_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/validator_subcommands.rs [hunk 8]
```diff
@@ -1,15 +1,17 @@
-use clap::Parser;
-use nimiq_keys::Address;
-use nimiq_rpc_interface::consensus::ConsensusInterface;
-use nimiq_rpc_interface::validator::ValidatorInterface;
+use std::sync::Arc;
 
 use anyhow::Error;
 use async_trait::async_trait;
+use clap::Parser;
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/validator_subcommands.rs]: We should be using `parking_lot`

File: rpc-client/src/subcommands/zkp_component_subcommands.rs [hunk 9]
```diff
@@ -1,11 +1,14 @@
+use std::sync::Arc;
+
 use anyhow::Error;
 use async_trait::async_trait;
 use clap::Parser;
-use nimiq_rpc_interface::zkp_component::ZKPComponentInterface;
+use tokio::sync::RwLock;
```
Comment (jsdanielh) [inline rpc-client/src/subcommands/zkp_component_subcommands.rs]: We should be using `parking_lot`

**Review by jsdanielh — CHANGES_REQUESTED** [review]
Need to update `nimiq-json-rpc` dependencies

File: rpc-client/src/main.rs [hunk 10]
```diff
@@ -1,8 +1,9 @@
 use anyhow::{bail, Error};
 use clap::Parser;
+
```
Comment (jsdanielh) [inline rpc-client/src/main.rs]: Remove extra line

File: rpc-client/src/main.rs [hunk 11]
```diff
@@ -102,9 +104,15 @@ impl Client {
...
             network: NetworkProxy::new(client.clone()),
-            zkp_component: ZKPComponentProxy::new(client),
+            zkp_component: ZKPComponentProxy::new(client.clone()),
+            ws_client: client,
         })
     }
+
+    /// Closes the WS connection
+    pub async fn close(&mut self) {
+        let _ = self.ws_client.close().await;
```
Comment (jsdanielh) [inline rpc-client/src/main.rs]: Remove clippy warning:
```suggestion
        self.ws_client.close().await;
```

**Review by jsdanielh — CHANGES_REQUESTED** [review]

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**